src/gurt/telemetry.c

```diff
@@ -2417,37 +2433,73 @@ d_tm_get_shared_memory(int srv_idx)
 			"%s\n", strerror(errno));
 		return NULL;
 	}
-	return addr;
+
+	rc = alloc_ctx(&new_ctx, addr);
+	if (rc != 0)
+		return NULL;
+
+	return new_ctx;
+}
+
+/**
+ * Detaches from a telemetry memory region and frees the context.
+ *
+ * \param[in]	ctx	Context to be freed
+ */
+void
+d_tm_close(struct d_tm_context **ctx)
+{
+	int			 rc;
+	struct d_tm_shmem_hdr	*shmem_root;
+
+	if (ctx == NULL || *ctx == NULL)
+		return;
+
+	shmem_root = (*ctx)->shmem_root;
+
+	if (shmem_root == NULL)
+		goto out;
+
+	rc = shmdt((void *)shmem_root);
+	if (rc < 0)
+		D_ERROR("Unable to detach from shared memory segment.  "
+			"shmdt failed, %s.\n", strerror(errno));
+out:
+	D_FREE(*ctx);
 }
 
 /**
  * Allocates memory from within the shared memory pool with 64-bit alignment
  * Clears the allocated buffer.
  *
+ * param[in]	shmem	The shmem pool in which to alloc
  * param[in]	length	Size in bytes of the region within the shared memory
  *			pool to allocate
  *
  * \return		Address of the allocated memory
  *			NULL if there was no more memory available
  */
-void *
-d_tm_shmalloc(int length)
+static void *
+d_tm_shmalloc(struct d_tm_shmem_hdr *shmem, int length)
 {
+	if (shmem == NULL || length == 0)
+		return NULL;
+
 	if (length % sizeof(uint64_t) != 0) {
 		length += sizeof(uint64_t);
 		length &= ~(sizeof(uint64_t) - 1);
 	}
 
-	if (d_tm_shmem_idx) {
-		if ((d_tm_shmem_free - length) > 0) {
-			d_tm_shmem_free -= length;
-			d_tm_shmem_idx += length;
-			D_DEBUG(DB_TRACE,
-				"Allocated %d bytes.  Now %" PRIu64 " remain\n",
-				length, d_tm_shmem_free);
-			memset((void *)(d_tm_shmem_idx - length), 0, length);
-			return d_tm_shmem_idx - length;
-		}
+	if (shmem->sh_bytes_free > 0 && length <= shmem->sh_bytes_free) {
```

Comment:
reverse the logic and return early on failure? then the rest of the logic can be dedented

Reply:
Another good point, will fix in a follow-on